File: android/app/src/main/java/com/rafaeltmbr/stopwatch/domain/use_cases/impl/SaveStopwatchStateUseCaseImpl.kt
package com.rafaeltmbr.stopwatch.domain.use_cases.impl

import com.rafaeltmbr.stopwatch.domain.data.repositories.StopwatchRepository
import com.rafaeltmbr.stopwatch.domain.data.stores.StateStore
import com.rafaeltmbr.stopwatch.domain.entities.StopwatchState
import com.rafaeltmbr.stopwatch.domain.use_cases.SaveStopwatchStateUseCase

class SaveStopwatchStateUseCaseImpl(
    private val store: StateStore<StopwatchState>,
    private val repository: StopwatchRepository,
) : SaveStopwatchStateUseCase {
    override suspend fun execute() {
        repository.save(store.state.value)
    }
}

File: ios/Stopwatch/core/use_cases/impl/SaveStopwatchStateUseCaseImpl.swift
class SaveStopwatchStateUseCaseImpl<MSS>: SaveStopwatchStateUseCase
where MSS: MutableStateStore, MSS.State == StopwatchState
{
    private(set) var stopwatchStore: MSS
    private(set) var stopwatchRepository: StopwatchRepository
    
    init(_ stopwatchState: MSS, _ stopwatchRepository: StopwatchRepository) {
        self.stopwatchStore = stopwatchState
        self.stopwatchRepository = stopwatchRepository
    }
    
    func execute() async {
        await stopwatchRepository.save(stopwatchStore.state)
    }
}
